### PR TITLE
Time series classifiers refactor/rocket classifier

### PR DESCRIPTION
**Pull request by victordremov**
#### Reference Issues/PRs

#1146 TSC refactor for `ROCKETEstimator`

#### What does this implement/fix? Explain your changes.

Refactor `ROCKETEstimator` class according to the [new extension template](https://github.com/alan-turing-institute/sktime/blob/main/extension_templates/classification.py).
- fit and predict turned private
- `capabilities` class attribute renamed to `_tags` class attribute
- capabilities names prefixed with `capability:` prefix
- new tags added to ESTIMATOR_TAG_REGISTER

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
- `capability:missing_values` and `capability:multivariate` are intentionally kept distinct from `handles-missing-values` and `univariate-only` to stick to `capability:` prefix convention at least until refactoring ends. It looks like it will be easier merge them after refactoring and rename all occurences at once.
- I do not know meaning for `train_estimate` and `contractable` tags, so I need help to fill them

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).

### CODE REVIEW

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [inline sktime/classification/kernel_based/_rocket_classifier.py:58]: Hello, Matthew,

Could you kindly help me to understand what `contractable` and `train_estimate` capabilities mean? I am refactoring ROCKERClassifier and I am not sure what they mean. @MatthewMiddlehurst

Comment (MatthewMiddlehurst) [reply sktime/classification/kernel_based/_rocket_classifier.py:58]: `contractable` The user can set a time limit for the fit process. i.e. in cBOSS instead of searching a set number of parameter sets, keep searching parameter sets until time runs out.

`train_estimate` The classifier can generate a set of estimated probabilites for the training dataset. i.e. in cBOSS again, cross-validation is used to determine how useful an ensemble member is. We can reuse this if an estimate of performance is required on the training set. See also: out-of-bag estimates for forest classifiers.

**Comment by fkiraly** [comment]
@victordremov, @MatthewMiddlehurst, could you kindly use the `_tags` system?
Then questions like this wouldn't arise.

In the `_tags` syste, we have the capability to attach explanations, e.g., if you use `registry.all_tags`.
Of course, you need to deposit the plain English description in `registry._tags` first for this to work.

Would you mind doing/updating that too, @MatthewMiddlehurst, @victordremov?

**Comment by fkiraly** [comment]
oh, I see you're already doing that - good work, thanks! Please ignore my comment.

File: sktime/classification/base.py
Comment (fkiraly) [inline sktime/classification/base.py:153]: I think you should be using `get_tag` here - `get_tag` does the same but also allows dynamic overrides.

File: sktime/classification/base.py
Comment (fkiraly) [inline sktime/classification/base.py:227]: 👍

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (fkiraly) [inline sktime/classification/kernel_based/_rocket_classifier.py:4]: the strings should be GitHub IDs.
I know it's inconsistent currently, but later we'll want to be using GitHub IDs.

File: sktime/registry/_tags.py
Comment (fkiraly) [inline sktime/registry/_tags.py:134]: that's not exactly true - it should mean "can classifier classify time series with 2 or more variables?"

File: sktime/registry/_tags.py
Comment (fkiraly) [inline sktime/registry/_tags.py:140]: better: "unequal length time series"

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (fkiraly) [inline sktime/classification/kernel_based/_rocket_classifier.py:104]: this should also go, it's in the base class too

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (fkiraly) [inline sktime/classification/kernel_based/_rocket_classifier.py:56]: I'd also put these in the base class with the following defaults:
```python
        "capability:multivariate": False,
        "capability:unequal_length": False,
        "capability:missing_values": False,
        "capability:train_estimate": False,
        "capability:contractable": False,
```
So here you would only have to put

```python
    _tags = {
        "capability:multivariate": True
    }
```

because that's the only one that would differ from the default - but no harm having all of them in the concrete class.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great work, @victordremov!

Change request with only very small things:

* `self._is_fitted = True` line from the class should also go
* `__author__` field should be populated with GitHub ID, the "real name" goes in the contributors file (so can be linked there via GitHub ID)
* some of the tags are not correctly described in the registry, I suggested plain English descriptions

Not a blocker if we don't do it, but possibly useful to do: putting defaults for the capability tags int he base classifier.

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (fkiraly) [reply sktime/classification/kernel_based/_rocket_classifier.py:58]: so for the descriptions, how about this:

contractable - `does the estimator support time limits on execution, aka contract time setting?`
train_estimate - `can the classifier produce out-of-bag predictions on the training set?`

Might be worth also adding an explicit mention to the method/argument if that's easy. What do you think, @MatthewMiddlehurst, on this as description strings?

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: I don't really like such implicit defaults, custom tag inheritance + defaults make things difficult. When all tags are listed in one class, it is okay, I can get it from the code. But if some tags are in base class, some in subclass - the only method to get tag values is to instantiate object and inspect it in runtime

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py:58]: looks like contractable is not self-explanatory enough, if nobody is against it, I would like to rename it into `capability:set-max-fit-time` after TSC refactoring completion and add TODO comment about it

File: sktime/classification/base.py
Comment (victordremov) [reply sktime/classification/base.py:153]: done 83a9faa1

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py]: 96c657bb

File: sktime/registry/_tags.py
Comment (victordremov) [reply sktime/registry/_tags.py]: bdd99a4a

File: sktime/registry/_tags.py
Comment (victordremov) [reply sktime/registry/_tags.py]: a0083bd9

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py]: ac47ae9c

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py:58]: For `contractable`, though, I think is better to emphasize that only train time can be limited, not predict, so `does the estimator support limiting max fit time?`
cbf10004

Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py:58]: I like `can the classifier produce out-of-bag predictions on the training set?` description for `train_estimate`
e94bec0a

**Comment by victordremov** [comment]
Hello, Franz,
Thank you very much for you review! I have gone through your suggestions, please look review the code again.

Note that I did not add default _tags values for BaseEstimator, I have wrote my reasons why we should not do it in https://github.com/alan-turing-institute/sktime/pull/1239#discussion_r684405424 cc @fkiraly

**Comment by TonyBagnall** [comment]
Hi, we call it contractable in all our papers and want to keep the name, sorry

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (fkiraly) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: >I don't really like such implicit defaults, custom tag inheritance + defaults make things difficult. When all tags are listed in one class, it is okay, I can get it from the code. But if some tags are in base class, some in subclass - the only method to get tag values is to instantiate object and inspect it in runtime

I get your point, and agree with it.

Comment (fkiraly) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: In fact, my opinion (in previous core developer discussions) was this:

* have default tags in the base class (not `BaseEstimator` but of the scitype, e.g., `BaseClassifier`), so we have "safe defaults" in case someone forgets
* also ask extenders to fill in all the tags in any concrete subclass, as a kind of "poka yoke" which makes it less likely that people forget to enter tags, and since it's easier to look at the code and see all the tags.

The consensus we ended up with though, was:

* have defaults in base classes
* do *not* ask to enter the tags which are equal to defaults

I don't like that entirely, but that's what we arrived at.

Comment (fkiraly) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: in any case, I wouldn't see it as a PR blocker if you fill in all the tags in the subclass.

Comment (TonyBagnall) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: well I was not in favour of custom tag inheritance, hence why I set up capabilities as a much simpler system. Personally I much prefer boolean tags to freeform strings. Sorry it be awkward, but these changes are going to effect every single classifier, so I would like a proper discussion of this rather than just slipping it through with a single classifier. We can refactor later if it is collectively decided to be desirable, but for now I would like a simple porting over to the _tags, and defaults in the base class as Franz suggests.  contractable can change to train-time-contractable if you want.

**Comment by TonyBagnall** [comment]
> * `__author__` field should be populated with GitHub ID, the "real name" goes in the contributors file (so can be linked there via GitHub ID)

is this policy? It is not done that way in most of the code. If so I'll enforce it, but was not aware

> * some of the tags are not correctly described in the registry, I suggested plain English descriptions

I can refine this once its in if you want.

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (MatthewMiddlehurst) [reply sktime/classification/kernel_based/_rocket_classifier.py:58]: Worth noting that while out-of-bag predictions are the best example, they aren't the only method of generating train set predictions. Maybe replace out-of-bag with bespoke?

Contracting description is good.

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (TonyBagnall) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: > In fact, my opinion (in previous core developer discussions) was this:
> 
> * have default tags in the base class (not `BaseEstimator` but of the scitype, e.g., `BaseClassifier`), so we have "safe defaults" in case someone forgets
> * also ask extenders to fill in all the tags in any concrete subclass, as a kind of "poka yoke" which makes it less likely that people forget to enter tags, and since it's easier to look at the code and see all the tags.
> 
> The consensus we ended up with though, was:
> 
> * have defaults in base classes
> * do _not_ ask to enter the tags which are equal to defaults
> 
> I don't like that entirely, but that's what we arrived at.

Oh ok. Having no defaults and also not insisting on assigning values will increase the chances of run time errors. My view would be to have scitype defaults but not insist on subclasses assigning a value, I view that as bloaty.  One point of inheritance is to provide the default behaviour or explicitly specialise. But as I say, this PR does not need to decide the whole issue, I suggest we just get this in with the minimal refactoring and then consider from there.

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py:58]: 1. If I understood Matthew right, `train_estimate` description should be generalized to cover other methods to estimate classifier performance. Changed it to `can the classifier estimate its performance on the training set?`. If there are better suggestions, will gladly change it. 4cbd4398
2. It looks like `contract time setting` is a synonym term for `contractable`. If so, better add it to description. `contract time setting, i.e. does the estimator support limiting max fit time?` e84da651

File: sktime/classification/kernel_based/_rocket_classifier.py
Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: I see there is still no agreement about defaults and overrides, and I agree with Tony we should keep things simple until there is some decision on this question.

So for now lets only rename capabilities to tags with `capability:` prefix, as it is very simple for contributors. If later we decide to add tags defaults to base class and enforce to remove from concrete classes tags that have default values, we can do it in automated way at any stage of refactoring or after it.

Comment (victordremov) [reply sktime/classification/kernel_based/_rocket_classifier.py:56]: I understand there is already an agreement to have default tag values, but it is not still clear where to keep them, in `BaseEstimator` or `BaseClassifier` and should concrete class list all tag values or only distinct from defaults. I guess when this decision is made, it should be put down in TSC template there https://github.com/alan-turing-institute/sktime/blob/84c4ff004b7e6735becc0e63f105709e2e0b3e8a/extension_templates/classification.py#L70

For now, lets just convert capabilities into tags.

**Comment by victordremov** [comment]
> Hi, we call it contractable in all our papers and want to keep the name, sorry

removed `TODO: rename tag` comment e84da651

**Comment by victordremov** [comment]
Hello, @fkiraly @TonyBagnall @MatthewMiddlehurst,

Fixed all the issues and waiting for your formal review.

File: sktime/classification/base.py
Comment (MatthewMiddlehurst) [reply sktime/classification/base.py:153]: predict_proba seems to still use `get_class_tag`. Very minor but a bit strange to use one in fit/predict and another in predict_proba unless theres a reason.

**Comment by fkiraly** [comment]
>is this policy? It is not done that way in most of the code. If so I'll enforce it, but was not aware

@TonyBagnall, I thought so? We really need that code style guide written, come doc sprint.

**Review by fkiraly — DISMISSED** [review]
can be merged from my side, nice work!

Thanks, @victordremov!

**Review by MatthewMiddlehurst — DISMISSED** [review]
From an earlier comment: `predict_proba` seems to still use `get_class_tag`. Very minor but a bit strange to use one in `fit`/`predict` and another in `predict_proba` unless theres a reason.

Other than that looks fine.

**Comment by fkiraly** [comment]
yes, agreed with @MatthewMiddlehurst 

There are now some new conflicts in the contributors file due to merging other branches, so if you could resolve those, plus the `get_class_tag` in `update`, we could merge right after that.

**Review by MatthewMiddlehurst — APPROVED** [review]

**Comment by dremovv** [comment]
> predict_proba seems to still use get_class_tag. Very minor but a bit strange to use one in fit/predict and another in predict_proba unless theres a reason.
@MatthewMiddlehurst Thank you for pointing this out. I was sure I have changed get_class_tag to get_tag in all 3 methods: fit, predict and predict_proba, but now I can see I have done only 2.

**Comment by MatthewMiddlehurst** [comment]
@dremovv I pushed some commits earlier which fixed the issue and the conflict, so no need for any action. There are a couple of high priority PRs that will go in before this one it seems, but I would like it get this in soon as it will impact some of my stuff.

Thanks for contributing.

**Review by fkiraly — APPROVED** [review]
great, good to go!

**Pull request merged**
